compiler/hash-layout/src/compute.rs

```diff
@@ -617,6 +617,55 @@ impl<'l> LayoutComputer<'l> {
         })
     }
 
+    /// Compute the range of values that the discriminants of an [Adt]. This
+    /// will compute the minimum and maximum values that the discriminant
+    /// can take.
+    fn compute_discriminant_range(
+        &self,
+        adt: &Adt,
+        variant_layouts: &IndexVec<VariantIdx, Vec<LayoutId>>,
+    ) -> (i128, i128) {
+        let (mut min, mut max) = (i128::MAX, i128::MIN);
+
+        let discr_ty = adt.discriminant_ty();
+        let bits = discr_ty.size(self.data_layout().pointer_size).bits();
+
+        // Now we iterate through the variants, and compute the range of the
+        // discriminant whilst skipping uninhabited variants.
+        for (variant, mut value) in adt.discriminants().map(|(v, discr)| (v, discr as i128)) {
+            let is_uninhabited =
+                self.store().map_many_fast(variant_layouts[variant].iter().copied(), |fields| {
+                    fields.iter().any(|f| f.abi.is_uninhabited())
+                });
+
+            if is_uninhabited {
+                continue;
+            }
+
+            if discr_ty.is_signed() {
+                value = (value << (128 - bits)) >> (128 - bits);
```

Comment:
Won't this be an off-by-one error? Why not use the raw `u128` each time?

Reply:
Well this is just a truncation of the value, where would the off-by-one come from?

Reply:
Well this is just a truncation of the value, where would the off-by-one come from?

Reply:
Why do you need to truncate the value?

Reply:
Strictly speaking, this is not required. However, when an `enum` value does overflow but it is not being used, i.e.
```rust
#[repr(i8)]
Item := enum(
  #[discriminant(1)]
  A
  #[discriminant(256)] // <-- True discriminant will be `-255`.
  B
)
```
So when it reaches layout computation, then the `256` value will be truncated to being the correct 8bits instead of being encoded in the raw `u128`.  

This shouldn't affect anything because this should already happen, I need to do some testing though before removing the actual check.

Reply:
Isn't that discriminant value invalid in the first place?

Reply:
Yes it would, I am just unsure if it requires truncation at any other point.

Reply:
I remembered what the problem was: later during the calculation of what the discriminant representation is, the range of the discriminant values is taken into account if `repr` isn't specifed. Additionally, there is a sanity check that the specified representation is enough to fit the value range (since this error is caught before the compilation gets to lowering). So, without this value truncation, the following check will go off https://github.com/hash-org/hashc/blob/323d483031ccc36b4e14894930f13162c1e4be0c/compiler/hash-ir/src/ty.rs#L994-L996

And this fails because of the way `fit_signed` works: https://github.com/hash-org/hashc/blob/323d483031ccc36b4e14894930f13162c1e4be0c/compiler/hash-target/src/abi.rs#L90-L100

This will be triggered for code like:
```rs
#[repr("i8")]
Direction := enum(
    #[discriminant(-120)]
    Up,
    Right,
    Down,
    Left,
)
```


So the truncation needs to happen or this will mis-categorise negative ints as always being `i128`s 

I will add a comment for this.